python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -308,3 +382,48 @@ def _inplace_subtract(node: ResourceDict, resources: ResourceDict) -> None:
         assert k in node, (k, node)
         node[k] -= v
         assert node[k] >= 0.0, (node, k, v)
+
+
+def _inplace_add(a: Dict, b: Dict) -> None:
+    for k, v in b.items():
+        a[k] += v
+
+
+def placement_groups_to_resource_demands(placement_group_load: List[
+        PlacementGroupTableData]):
+    """Preprocess placement group requests into regular resource demand vectors
+    when possible.
+
+    Args:
+        placement_group_load (List[PlacementGroupData]): List of
+        PlacementGroupLoad's.
+
+    Returns:
+        (resource_demand_vector: List[ResourceDict], unconverted:
+        List[List[ResourceDict]]) Converts PACK, SPREAD, and STRICT_PACK
+        placement groups to resource demands. STRICT_SPREAD is not equivalent
+        to a resource demand so it can't be converted.
+
+    """
+    resource_demand_vector = []
+    unconverted = []
+    for placement_group in placement_group_load:
+        shapes = [dict(bundle.unit_resources) for bundle in placement_group.bundles]
+        if (placement_group.strategy == PlacementStrategy.PACK or
+                placement_group.strategy == PlacementStrategy.SPREAD):
+            resource_demand_vector.extend(shapes)
+        elif placement_group.strategy == PlacementStrategy.STRICT_PACK:
+            combined = collections.defaultdict(float)
+            for shape in shapes:
+                for label, quantity in shape.items():
+                    combined[label] += quantity
+            resource_demand_vector.append(combined)
+        elif (placement_group.strategy ==
+              PlacementStrategy.STRICT_SPREAD):
+            unconverted.append(shapes)
+        else:
+            logger.error(
```

Comment:
shouldn't we use the cli_logger.error?

Reply:
No, autoscaler output shouldn't be written to the cli.